### PR TITLE
SHA512/224 and SHA512/256

### PR DESCRIPTION
**Pull request by paulidale**
Support added for these two digests, available only via the EVP interface.

- [x] tests are added or updated

### CODE REVIEW

**Comment by bernd-edlinger** [comment]
documentation?

File: crypto/include/internal/sha.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ /* Copyright (c) 2017 National Security Research Institute.  All rights reserved. */
+
+#ifndef HEADER_INTERNAL_SHA_H
+# define HEADER_INTERNAL_SHA_H
+
+# include <openssl/opensslconf.h>
+
+# ifdef  __cplusplus
+extern "C" {
+# endif
```
Comment (dot-asm) [inline crypto/include/internal/sha.h]: What's the point of extern "C" in *internal* header. Users won't have opportunity to compile it with C++ compiler...

File: test/recipes/30-test_evp_data/evpdigest.txt [hunk 2]
```diff
@@ -167,6 +167,62 @@ Input = "a"
 Ncopy = 64
 Output = e718483d0ce769644e2e42c7bc15b4638e1f98b13b2044285632a803afa973ebde0ff244877ea60a4cb0432ce577c31beb009c5c2c49aa2e4eadb217ad8cc09b
 
+Digest = SHA512-224
+Input = 
+Output = 6ed0dd02806fa89e25de060c19d3ac86cabb87d6a0ddd05c333b84f4
```
Comment (dot-asm) [inline test/recipes/30-test_evp_data/evpdigest.txt]: Reference to source would be more than appropriate. I mean where the test vectors come from.

File: crypto/sha/sha512.c [hunk 3]
```diff
@@ -59,6 +60,42 @@
 # define SHA512_BLOCK_CAN_MANAGE_UNALIGNED_DATA
 #endif
 
+int sha512_224_init(SHA512_CTX *c)
```
Comment (dot-asm) [inline crypto/sha/sha512.c]: I don't recall that lower-case would be reflection of symbol visibility... I mean I don't think there is a rule that says that unexported symbols **has to be*** lower case...

File: crypto/include/internal/sha.h (hunk 1 above)
Comment (paulidale) [reply crypto/include/internal/sha.h]: Good point.  I'll fix.

File: test/recipes/30-test_evp_data/evpdigest.txt (hunk 2 above)
Comment (paulidale) [reply test/recipes/30-test_evp_data/evpdigest.txt]: The vectors are from NIST, I'll add references tomorrow.

File: crypto/sha/sha512.c (hunk 3 above)
Comment (paulidale) [reply crypto/sha/sha512.c]: I was in a bit of a quandary over this and other nuances.

The move seems to be towards not exporting lower level functions, thus I changed the the names to lowercase which seems to (mostly) be the convention.  I also didn't export the evp_sha512_224() function, instead only allowing access via the string representation.

I'm happy to change either or to export everything but we'll want consensus first.

**Comment by paulidale** [comment]
@bernd-edlinger I'm not sure where or what to document.  There are no exported functions here and the documentation pages deal with these.  Suggestions are welcome.

I'm happy to add something to the CHANGES file if that's deemed appropriate.  I feel that it would be but not strongly.

File: crypto/sha/sha512.c (hunk 3 above)
Comment (richsalz) [reply crypto/sha/sha512.c]: I think the EVP function should be available; the others are.  I think all-lowercase-names for internal functions is common, even if not a formal standard of ours so I would stick with that.

**Comment by richsalz** [comment]
If you make the EVP API public, then it could go in one of the EVP_sha manpages.

**Comment by paulidale** [comment]
Fixed things as per comments.

Added test sources, removed the C++ wrapper, made the EVP_sha512_224 and EVP_sha512_256 functions public and added documentation.

**Comment by paulidale** [comment]
Comments, suggestions or approval anyone?

**Comment by paulidale** [comment]
@dot-asm ?

**Comment by paulidale** [comment]
Rebased to trunk -- the only conflict being in the `libcrypto.num` file.

**Comment by dot-asm** [comment]
> @dot-asm ?

I don't quite follow what is expected from me. My approval doesn't make difference in another committer's request, so in cases when review outcome is insignificant as a rule I'm reluctant to add one. As in this one, just nits...

**Comment by paulidale** [comment]
Since you'd reviewed already, I figured it would be polite to get your confirmation as well.

**Comment by paulidale** [comment]
Merged, thanks.

**Pull request closed without merging**